aiohttp/client.py

```diff
@@ -761,6 +761,10 @@ async def _connect_and_send_request(
                             data = None
                             if headers.get(hdrs.CONTENT_LENGTH):
                                 headers.pop(hdrs.CONTENT_LENGTH)
+                        elif resp.status in (307, 308) and req._body is not None:
```

Comment:
Should we just make this else? Seems like it should still happen on (e.g.) 302 that is not POST. i.e. We should either change to GET method, or reuse the payload.

Reply:
Are there any other codes that allow preserving the payload? I'm worried about sending the payload when we shouldn't

Reply:
Anything that doesn't match the first if condition is going to reuse the data and method.

Reply:
let me add that

Reply:
changed it to else: and wrote tests to make sure we don't leak it and the 301/302 case works as expected